### PR TITLE
Factor out repeated code in main.

### PR DESCRIPTION
**Pull request by snim2**
This is a small improvement in `main()`, but I've had to use the `unreachable!()` macro to avoid errors like this:

```rust
error[E0308]: match arms have incompatible types
   --> src/main.rs:153:22
    |
153 |       let extension2 = match Path::new(&args.arg_diff_file).extension() {
    |  ______________________^
154 | |         Some(ext) => ext.to_str().unwrap(),
155 | |         None => {
156 | |             exit_on_err(format!("Cannot determine file type of {}.", args.arg_diff_file));
157 | |             // unreachable!()
158 | |         }
159 | |     };
    | |_____^ expected &str, found ()
    |
    = note: expected type `&str`
               found type `()`
```

Is there a better way?

### CODE REVIEW

**Comment by jaheba** [comment]
```rust
fn exit_on_err(message: String) {
// to
fn exit_on_err(message: String) -> ! {
```

basically means that exit_on_err does not return

**Comment by snim2** [comment]
Oh, that's really neat!

File: src/main.rs [hunk 1]
```diff
@@ -103,10 +107,7 @@ fn parse_file(filename: &str, lexer_path: &str, yacc_path: &str) -> ast::Arena<S
     };
     ast::parse_file(filename)
         .map_err(error_to_str)
-        .map_err(|msg| {
-                     writeln!(&mut stderr(), "{}", msg).ok();
-                     process::exit(1);
-                 })
+        .map_err(|msg| { exit_on_err(msg); })
```
Comment (jaheba) [inline src/main.rs]: AFAIK this could just be `.map_err(exit_on_err)`.

File: src/main.rs [hunk 2]
```diff
@@ -77,15 +77,19 @@ struct Args {
     flag_version: bool,
 }
 
+fn exit_on_err(message: String) -> ! {
```
Comment (jaheba) [inline src/main.rs]: I think, you normally would just want to have `message: &str` as the parameter, but that requires the function to be called like this

```rust
exit_on_err(&format!("Could not {} file {}.", action, filepath));
```
Which looks a bit more ugly.

**Comment by snim2** [comment]
Well, it seems we can have one of these changes but not both ;)

**Comment by jaheba** [comment]
That's why we can't have nice things :/

I believe using `ref` instead of `&` is generally a bit more defensive, because that closure can take both owned and borrowed values. I can be totally wrong though.

```rust
        .map_err(|ref msg| exit_on_err(msg) )
```

At least I would remove the block (`{ .. }`)  so either

```rust
        .map_err(|ref msg| exit_on_err(msg) )
```
or

```rust
        .map_err(|msg| exit_on_err(&msg) )
```

should do the job.

**Comment by snim2** [comment]
Good to squash?

**Comment by jaheba** [comment]
Looks good to me. Maybe we can find a better name then `exit_on_err`, because for me it implies that the function checks, whether there might an error. Maybe `exit_with_err` is a bit better, but I'm not too sure either.

**Comment by jaheba** [comment]
Even better :) If you squash, I will merge.

**Comment by snim2** [comment]
Squashed!

**Comment by snim2** [comment]
Thanks!

**Pull request merged**